File: services/ordering/src/main/java/io/wkrzywiec/fooddelivery/ordering/domain/Order.java
package io.wkrzywiec.fooddelivery.ordering.domain;

import io.wkrzywiec.fooddelivery.commons.infra.store.DomainEvent;
import io.wkrzywiec.fooddelivery.commons.model.CreateOrder;
import io.wkrzywiec.fooddelivery.commons.infra.messaging.IntegrationMessage;
import io.wkrzywiec.fooddelivery.ordering.domain.outgoing.*;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

import java.math.BigDecimal;
import java.util.*;

import static io.wkrzywiec.fooddelivery.ordering.domain.OrderStatus.*;
import static java.lang.String.format;

@Getter
@EqualsAndHashCode
@ToString
public class Order {
    private String id;
    private int version;
    private String customerId;
    private String farmId;
    private OrderStatus status;
    private String address;
    private List<Item> items;
    private BigDecimal deliveryCharge;
    private BigDecimal tip = new BigDecimal(0);
    private BigDecimal total = new BigDecimal(0);

    List<DomainEvent> changes = new ArrayList<>();

    private Order() {}

    private Order(String id, String customerId, String farmId, List<Item> items, String address, BigDecimal deliveryCharge) {
        this(id, customerId, farmId, CREATED, address, items, deliveryCharge, BigDecimal.ZERO);
    }

    private Order(String id, String customerId, String farmId, OrderStatus status, String address, List<Item> items, BigDecimal deliveryCharge, BigDecimal tip) {
        if (id == null) {
            this.id = UUID.randomUUID().toString();
        } else {
            this.id = id;
        }
        this.customerId = customerId;
        this.farmId = farmId;
        this.status = status;
        this.address = address;
        this.items = items;
        this.deliveryCharge = deliveryCharge;
        this.tip = tip;
        this.calculateTotal();
    }

    static Order from(CreateOrder createOrder) {
        var order = new Order(
                createOrder.orderId(),
                createOrder.customerId(),
                createOrder.farmId(),
                mapItems(createOrder.items()),
                createOrder.address(),
                createOrder.deliveryCharge());

        order.changes.add(
                new OrderingEvent.OrderCreated(
                        order.id, 0,
                        order.customerId, order.farmId,
                        order.address, order.items,
                        order.deliveryCharge, order.total
                )
        );
        return order;
    }

    private static List<Item> mapItems(List<io.wkrzywiec.fooddelivery.commons.model.Item> items) {
        return items.stream().map(dto -> Item.builder()
                .name(dto.name())
                .amount(dto.amount())
                .pricePerItem(dto.pricePerItem())
                .build()).toList();
    }

    static Order from(List<OrderingEvent> events) {
        Order order = null;
        for (OrderingEvent event: events) {
            switch (event) {
                case OrderingEvent.OrderCreated created -> {
                    order = new Order();

                    order.id = created.orderId();
                    order.version = 0;
                    order.status = CREATED;
                    order.customerId = created.customerId();
                    order.farmId = created.farmId();
                    order.items = created.items();
                    order.address = created.address();
                    order.deliveryCharge = created.deliveryCharge();
                    order.tip = BigDecimal.ZERO;
                }

                case OrderingEvent.OrderCanceled canceled -> {
                    order.status = CANCELED;
                    order.version = canceled.version();
                }

                case OrderingEvent.OrderInProgress inProgress -> {
                    order.status = IN_PROGRESS;
                    order.version = inProgress.version();
                }

                case OrderingEvent.TipAddedToOrder tipAdded -> {
                    order.tip = tipAdded.tip();
                    order.version = tipAdded.version();
                }

                case OrderingEvent.OrderCompleted completed -> {
                    order.status = COMPLETED;
                    order.version = completed.version();
                }

                default -> throw new IllegalArgumentException("Failed to replay events to build order object. Unhandled events: " + event.getClass());
            }
        }
        return order;
    }

    List<DomainEvent> uncommittedChanges() {
        return changes;
    }

    void calculateTotal() {
        this.total = items.stream()
                .map(item -> item.getPricePerItem().multiply(BigDecimal.valueOf(item.getAmount())))
                .reduce(BigDecimal.ZERO, BigDecimal::add)
                .add(deliveryCharge)
                .add(tip);
    }

    void cancelOrder(String reason) {
        if (status != OrderStatus.CREATED) {
            throw new OrderingException(format("Failed to cancel an %s order. It's not possible to cancel an order with '%s' status", id, status));
        }
        this.status = CANCELED;

        increaseVersion();
        changes.add(new OrderingEvent.OrderCanceled(id, version, reason));
    }

    void setInProgress() {
        if (status == CREATED) {
            this.status = IN_PROGRESS;

            increaseVersion();
            changes.add(new OrderingEvent.OrderInProgress(id, version));
            return;
        }
        throw new OrderingException(format("Failed to set an '%s' order to IN_PROGRESS. It's not allowed to do it for an order with '%s' status", id, status));
    }

    public void addTip(BigDecimal tip) {
        this.tip = tip;
        this.calculateTotal();

        increaseVersion();
        changes.add(new OrderingEvent.TipAddedToOrder(id, version, tip, total));
    }

    public void complete() {
        if (status == IN_PROGRESS) {
            this.status = COMPLETED;

            increaseVersion();
            changes.add(new OrderingEvent.OrderCompleted(id, version));
            return;
        }
        throw new OrderingException(format("Failed to set an '%s' order to COMPLETED. It's not allowed to do it for an order with '%s' status", id, status));
    }

    private void increaseVersion() {
        this.version = this.version + 1;
    }
}

File: services/ordering/src/main/kotlin/io/wkrzywiec/fooddelivery/ordering/domain/Order.kt
package io.wkrzywiec.fooddelivery.ordering.domain

import io.wkrzywiec.fooddelivery.commons.infra.store.DomainEvent
import io.wkrzywiec.fooddelivery.commons.model.CreateOrder
import java.math.BigDecimal
import java.util.*


class Order private constructor(
    val id: UUID = UUID.randomUUID(),
    version: Int,
    val customerId: String,
    val farmId: String,
    status: OrderStatus = OrderStatus.CREATED,
    val address: String,
    val items: List<Item>,
    val deliveryCharge: BigDecimal,
    tip: BigDecimal
) {

    var version: Int = version
        private set
    var status: OrderStatus = status
        private set
    var tip: BigDecimal = tip
        private set
    var total: BigDecimal = BigDecimal(0)
        private set

    private val changes: MutableList<DomainEvent> = mutableListOf()

    init {
        total = calculateTotal()
    }

    private fun calculateTotal(): BigDecimal {
        return items.sumOf { it.pricePerItem.multiply(BigDecimal(it.amount)) }
            .add(deliveryCharge)
            .add(tip)
    }

    fun uncommittedChanges(): List<DomainEvent> {
        return changes
    }

    fun cancelOrder(reason: String) {
        if (status != OrderStatus.CREATED) {
            throw OrderingException("Failed to cancel an $id order. It's not possible to cancel an order with '$status' status")
        }
        status = OrderStatus.CANCELED

        increaseVersion()
        changes.add(OrderingEvent.OrderCanceled(id, version, reason))
    }

    fun setInProgress() {
        if (status == OrderStatus.CREATED) {
            status = OrderStatus.IN_PROGRESS

            increaseVersion()
            changes.add(OrderingEvent.OrderInProgress(id, version))
            return
        }
        throw OrderingException(
            "Failed to set an '$id' order to IN_PROGRESS. It's not allowed to do it for an order with '$status' status"
        )
    }

    fun addTip(tip: BigDecimal) {
        this.tip = tip
        total = calculateTotal()

        increaseVersion()
        changes.add(OrderingEvent.TipAddedToOrder(id, version, tip, total))
    }

    fun complete() {
        if (status == OrderStatus.IN_PROGRESS) {
            status = OrderStatus.COMPLETED

            increaseVersion()
            changes.add(OrderingEvent.OrderCompleted(id, version))
            return
        }
        throw OrderingException(
            "Failed to set an '$id' order to COMPLETED. It's not allowed to do it for an order with $status' status"
        )
    }

    private fun increaseVersion() {
        version += 1
    }

    companion object {

        @JvmStatic
        fun from(createOrder: CreateOrder): Order {
            val order = Order(
                id = createOrder.orderId,
                version = 0,
                customerId = createOrder.customerId,
                farmId = createOrder.farmId,
                status = OrderStatus.CREATED,
                address = createOrder.address,
                items = mapItems(createOrder.items),
                deliveryCharge = createOrder.deliveryCharge,
                tip = BigDecimal.ZERO
            )
            order.changes.add(OrderingEvent.OrderCreated(
                order.id, 0,
                order.customerId, order.farmId,
                order.address, order.items,
                order.deliveryCharge, order.total
            ))
            return order
        }

        private fun mapItems(items: List<io.wkrzywiec.fooddelivery.commons.model.Item>): List<Item> {
            return items.stream()
                .map{
                    Item(name = it.name, amount = it.amount, pricePerItem = it.pricePerItem)
                }
                .toList()
        }

        @JvmStatic
        fun from(events: List<OrderingEvent>): Order {
            lateinit var order: Order
            for (event in events) {
                when (event) {
                    is OrderingEvent.OrderCreated -> {
                        order = Order(
                            id = event.orderId,
                            version = 0,
                            customerId = event.customerId,
                            farmId = event.farmId,
                            status = OrderStatus.CREATED,
                            address = event.address,
                            items = event.items,
                            deliveryCharge = event.deliveryCharge,
                            tip = BigDecimal.ZERO
                        )
                    }

                    is OrderingEvent.OrderCanceled -> {
                        order.status = OrderStatus.CANCELED
                        order.version = event.version
                    }

                    is OrderingEvent.OrderInProgress -> {
                        order.status = OrderStatus.IN_PROGRESS
                        order.version = event.version
                    }

                    is OrderingEvent.TipAddedToOrder -> {
                        order.tip = event.tip
                        order.version = event.version
                    }

                    is OrderingEvent.OrderCompleted -> {
                        order.status = OrderStatus.COMPLETED
                        order.version = event.version
                    }

                    else -> throw IllegalArgumentException("Failed to replay events to build order object. Unhandled events: ${event::class.simpleName}")
                }
            }
            return order
            }
    }
}
